File: src/core/interfaces.ts
/**
 * Core interfaces for the message interception system
 */

import { 
  Message, 
  WhatsAppMessage, 
  RoutingDecision, 
  ConversationContext,
  AgentPersona,
  AIEngineConfig,
  ConversationInitiationRequest
} from './types';

/**
 * Interface for message routing logic
 */
export interface IMessageRouter {
  route(message: Message): Promise<RoutingDecision>;
  registerRule(name: string, rule: RoutingRule): void;
  removeRule(name: string): void;
  setDefaultAction(action: RoutingDecision): void;
}

/**
 * Routing rule definition
 */
export interface RoutingRule {
  name: string;
  priority: number;
  condition: (message: Message) => boolean | Promise<boolean>;
  action: (message: Message) => RoutingDecision | Promise<RoutingDecision>;
  enabled?: boolean;
}

/**
 * Interface for conversation management
 */
export interface IConversationManager {
  createConversation(
    phoneNumber: string,
    initialMessage?: Message,
    metadata?: Record<string, any>
  ): Promise<ConversationContext>;
  
  getConversation(conversationId: string): Promise<ConversationContext | null>;
  
  getActiveConversationByPhone(phoneNumber: string): Promise<ConversationContext | null>;
  
  updateConversation(
    conversationId: string,
    message: Message
  ): Promise<void>;
  
  closeConversation(conversationId: string): Promise<void>;
  
  listActiveConversations(filters?: {
    agentId?: string;
    phoneNumber?: string;
    state?: string;
  }): Promise<ConversationContext[]>;
  
  getConversationHistory(
    conversationId: string,
    limit?: number,
    offset?: number
  ): Promise<Message[]>;
}

/**
 * Interface for storage backend
 */
export interface IStorageBackend {
  // Key-value operations
  get<T = any>(key: string): Promise<T | null>;
  set<T = any>(key: string, value: T, ttl?: number): Promise<void>;
  delete(key: string): Promise<void>;
  exists(key: string): Promise<boolean>;
  
  // List operations
  listPush<T = any>(key: string, value: T): Promise<void>;
  listRange<T = any>(key: string, start: number, stop: number): Promise<T[]>;
  listLength(key: string): Promise<number>;
  
  // Set operations
  setAdd(key: string, member: string): Promise<void>;
  setMembers(key: string): Promise<string[]>;
  setRemove(key: string, member: string): Promise<void>;
  
  // Hash operations
  hashSet(key: string, field: string, value: any): Promise<void>;
  hashGet(key: string, field: string): Promise<any>;
  hashGetAll(key: string): Promise<Record<string, any>>;
  
  // Pattern operations
  keys(pattern: string): Promise<string[]>;
  
  // Transactions
  multi(): IStorageTransaction;
}

/**
 * Storage transaction interface
 */
export interface IStorageTransaction {
  get(key: string): IStorageTransaction;
  set(key: string, value: any, ttl?: number): IStorageTransaction;
  delete(key: string): IStorageTransaction;
  exec(): Promise<any[]>;
  discard(): void;
}

/**
 * Interface for AI engines
 */
export interface IAIEngine {
  generateResponse(
    message: Message,
    conversationHistory: Message[],
    context: {
      persona?: AgentPersona;
      conversationContext?: ConversationContext;
      customPrompt?: string;
    }
  ): Promise<string>;
  
  getCapabilities(): {
    supportsStreaming: boolean;
    maxContextLength: number;
    supportsFunctionCalling: boolean;
    supportsVision: boolean;
  };
  
  getName(): string;
}

/**
 * Interface for AI agents
 */
export interface IAgent {
  id: string;
  persona: AgentPersona;
  
  processMessage(
    message: Message,
    conversation: ConversationContext
  ): Promise<Message>;
  
  setEngine(engine: IAIEngine): void;
  getEngine(): IAIEngine | null;
  
  canHandle(message: Message): boolean;
}

/**
 * Interface for message interceptor
 */
export interface IMessageInterceptor {
  intercept(message: WhatsAppMessage): Promise<Message | null>;
  
  registerPreProcessor(processor: MessageProcessor): void;
  registerPostProcessor(processor: MessageProcessor): void;
  
  registerAgent(agent: IAgent): void;
  getAgent(agentId: string): IAgent | undefined;
  
  startProcessing(): Promise<void>;
  stopProcessing(): Promise<void>;
}

/**
 * Message processor function type
 */
export type MessageProcessor = (message: Message) => Promise<Message> | Message;

/**
 * WhatsApp API interface for sending messages
 */
export interface IWhatsAppAPI {
  sendMessage(
    to: string,
    message: string,
    options?: {
      replyTo?: string;
      buttons?: Array<{ id: string; title: string }>;
      mediaUrl?: string;
      mediaType?: 'image' | 'video' | 'document';
    }
  ): Promise<{ messageId: string; status: string }>;
  
  markAsRead(messageId: string): Promise<void>;
  
  getMessageStatus(messageId: string): Promise<string>;
}

/**
 * Pub/Sub extension for conversation initiation
 */
export interface IPubSubExtension {
  publishToTopic(
    topicId: string,
    message: string,
    options?: {
      initiateConversation?: boolean;
      conversationConfig?: Partial<ConversationInitiationRequest>;
    }
  ): Promise<void>;
  
  getTopicSubscribers(topicId: string): Promise<string[]>;
  
  initiateConversations(
    request: ConversationInitiationRequest
  ): Promise<{ 
    initiated: string[]; 
    failed: Array<{ phoneNumber: string; error: string }> 
  }>;
}

File: message_system/core/interfaces.py
"""
Core interfaces and abstract base classes
"""
from abc import ABC, abstractmethod
from typing import Optional, Dict, Any, List, Callable
from .message import Message, ConversationContext


class RoutingDecision:
    """Represents a routing decision for a message"""
    
    def __init__(self, 
                 should_process: bool,
                 conversation_id: Optional[str] = None,
                 create_new_conversation: bool = False,
                 reason: str = "",
                 metadata: Dict[str, Any] = None):
        self.should_process = should_process
        self.conversation_id = conversation_id
        self.create_new_conversation = create_new_conversation
        self.reason = reason
        self.metadata = metadata or {}


class IMessageRouter(ABC):
    """Interface for message routing logic"""
    
    @abstractmethod
    async def route(self, message: Message) -> RoutingDecision:
        """
        Determine how to route an incoming message
        
        Returns:
            RoutingDecision object containing routing information
        """
        pass
    
    @abstractmethod
    async def register_rule(self, name: str, rule: Callable) -> None:
        """Register a custom routing rule"""
        pass


class IConversationManager(ABC):
    """Interface for conversation management"""
    
    @abstractmethod
    async def create_conversation(self, 
                                 initial_message: Message,
                                 metadata: Dict[str, Any] = None) -> ConversationContext:
        """Create a new conversation"""
        pass
    
    @abstractmethod
    async def get_conversation(self, conversation_id: str) -> Optional[ConversationContext]:
        """Retrieve a conversation by ID"""
        pass
    
    @abstractmethod
    async def update_conversation(self, 
                                 conversation_id: str,
                                 message: Message) -> None:
        """Update conversation with new message"""
        pass
    
    @abstractmethod
    async def list_active_conversations(self, 
                                       participant_id: Optional[str] = None) -> List[ConversationContext]:
        """List active conversations, optionally filtered by participant"""
        pass
    
    @abstractmethod
    async def close_conversation(self, conversation_id: str) -> None:
        """Mark a conversation as closed"""
        pass


class IStorageBackend(ABC):
    """Interface for storage backend (can be implemented with files, Redis, etc.)"""
    
    @abstractmethod
    async def get(self, key: str) -> Optional[Any]:
        """Get value by key"""
        pass
    
    @abstractmethod
    async def set(self, key: str, value: Any, ttl: Optional[int] = None) -> None:
        """Set value with optional TTL"""
        pass
    
    @abstractmethod
    async def delete(self, key: str) -> None:
        """Delete value by key"""
        pass
    
    @abstractmethod
    async def exists(self, key: str) -> bool:
        """Check if key exists"""
        pass
    
    @abstractmethod
    async def list_keys(self, pattern: str) -> List[str]:
        """List keys matching pattern"""
        pass
    
    @abstractmethod
    async def append_to_list(self, key: str, value: Any) -> None:
        """Append value to a list"""
        pass
    
    @abstractmethod
    async def get_list(self, key: str, start: int = 0, end: int = -1) -> List[Any]:
        """Get list values"""
        pass


class IAgentEngine(ABC):
    """Interface for AI agent engines (ChatGPT, Claude, Gemini, etc.)"""
    
    @abstractmethod
    async def generate_response(self, 
                               message: Message,
                               conversation_history: List[Message],
                               context: Dict[str, Any] = None) -> Message:
        """Generate a response to a message"""
        pass
    
    @abstractmethod
    def get_engine_name(self) -> str:
        """Get the name of the engine"""
        pass
    
    @abstractmethod
    def get_capabilities(self) -> Dict[str, Any]:
        """Get engine capabilities"""
        pass


class IAgent(ABC):
    """Interface for AI agents with personas"""
    
    @abstractmethod
    async def process_message(self, 
                             message: Message,
                             conversation_context: ConversationContext) -> Message:
        """Process a message and generate response"""
        pass
    
    @abstractmethod
    def get_persona(self) -> Dict[str, Any]:
        """Get agent persona configuration"""
        pass
    
    @abstractmethod
    def set_engine(self, engine: IAgentEngine) -> None:
        """Set the AI engine for this agent"""
        pass